alerters/alerters-plugins/alerters-gnocchi/src/main/java/org/hawkular/alerter/gnocchi/GnocchiQuery.java

```diff
@@ -0,0 +1,391 @@
+                    Map<String, String> metrics = (Map<String, String>) resource.get("metrics");
+                    for (Map.Entry<String, String> metric : metrics.entrySet()) {
+                        Map<String, String> resultMetric = new HashMap<>();
+                        resultMetric.put("name", metric.getKey());
+                        resultMetric.put("id", metric.getValue());
+                        rawMetrics.add(resultMetric);
+                    }
+                }
+            }
+            log.debugf("Gnocchi Metrics found %s", rawMetrics);
+            conn.disconnect();
+            return rawMetrics;
+        } catch (IOException e) {
+            log.errorf(e,"Error querying Gnocchi resources %s", resourcesUrl);
+        }
+        return Collections.EMPTY_LIST;
+    }
+
+    private long intervalStart() {
+        int value = getIntervalValue(interval);
+        switch (getIntervalUnit(interval)) {
+            case HOURS:
+                value = value * 3600 * 1000;
+                break;
+            case MINUTES:
+                value = value * 60 * 1000;
+                break;
+            case SECONDS:
+                value = value * 1000;
+                break;
+        }
+        return (long)((System.currentTimeMillis() - value) / 1000);
+    }
+
+    public long parseTimestamp(String timestamp) {
+        try {
+            return ZonedDateTime.parse(timestamp, DateTimeFormatter.ISO_OFFSET_DATE_TIME).toInstant().toEpochMilli();
+        } catch (Exception e) {
+            log.debugf("Not able to parse [%s] with format [%s]", timestamp, DateTimeFormatter.ISO_OFFSET_DATE_TIME);
+        }
+        return -1;
+    }
+
+    public void run() {
+        List<Future> futures = new ArrayList<>();
+        for (Map.Entry<String, String> query : queries.entrySet()) {
+            futures.add(executorService.submit(() -> {
+                String metricName = query.getKey();
+                String measuresUrl = query.getValue() + "start=" + intervalStart();
+                try {
+                    URL url = new URL(measuresUrl);
+                    HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+                    conn.setRequestProperty(AUTHORIZATION, basicAuth);
+                    conn.setRequestMethod(GET);
+                    conn.setDoInput(true);
+                    List measures = JsonUtil.getMapper().readValue(conn.getInputStream(), List.class);
+                    List<Data> data = new ArrayList<>();
+                    for (int i = 0; i < measures.size(); i++) {
+                        List measure = (List)measures.get(i);
+                        if (measure != null && measure.size() == 3) {
+                            String timestamp = (String)measure.get(0);
+                            Double value = (Double)measure.get(2);
+                            data.add(Data.forNumeric(trigger.getTenantId(), metricName, parseTimestamp(timestamp), value));
```

Comment:
i think there is something missing here (or when you're building the queries). in gnocchi, if you dont' specify a granularity (http://gnocchi.xyz/operating.html#default-archive-policies), it will return back all granularities associated with metric. if you're using 'low' archive policy, this is only one granularity and you're fine. but if you're using 'medium', 'high', and presumably any custom policy, it will have multiple granularities returned and if you ignore it or don't specify it, you'll be effectively squashing multiple series into one.

Reply:
Yeah, in this first draft I'm taking the default granularity.
So effectively ignored.
I think we can add the granularity as part of the configuration (and if not present then fall into the default one).
I think yes, but in Gnocchi is a normal task to ask for same granularity in a group of metrics ?
i.e.
The example there there are 8 metrics for cpu, cpu-.@cpu-user-0, I guess that we can use the same granularity for all metrics fetched.

Or, is there another way where gnocchi users are familar ? I would like that the integration will be most natural for them (in the context that they need to define them).

Reply:
hmm... there really isn't a default granularity. when you create a metric, you must specify an archive policy (if i remember correctly). if the metric is a 'low' archive policy, then there is only one granularity (a timeseries of 5min aggregates over 30 days) and you have a 'default' granularity. but if they created a metric using another policy such as 'medium', then there will be multiple time-series (captures one series at 1min aggregates for 7 days, and another series at 1hr aggregates for a year). in the 'medium' case, if you get_measures without a granularity,  it will return both series, which will give you two different, partially overlapping (at different granularities), series.

for example, if you send in a point for 2017-01-01T00:05:32 with value 5. you will get back something roughly like:

`[(2017-01-01T00:00:00, 3600, 5),
  (2017-01-01T00:05:00, 60, 5)]`

two different series in a single result

Reply:
Good point, I will adapt the alert to take this case.

Reply:
That's interesting, so you can get potentially multiple running aggregates depending on the archive policy.  @lucasponce, I guess we either need to limit to one result or handle multiple results in a sane way.

Reply:
@jshaughn yes, that is was I try to comment.
The alerer takes granularity by default thinking that only one result was given but as multiple can be returnerd, granularity needs to be explicit in the trigger configuration to get a single result.